### PR TITLE
chore: configure golangci-lint

### PR DESCRIPTION
**Pull request by sudo-suhas**
- There are two versions of the configuration file added - .golangci.toml and .golangci-prod.toml. The former is for local development and testing, while the latter is for continuous integration (CI). The difference is that the prod version includes some more intensive checks that may take longer to run, but are important for ensuring code quality and security.
- Manage tools used to build, lint and format code using tools/go.mod. Set up the Makefile to use locally built versions of the tools using tools/go.mod.
- Update the golang-lint-action workflow to use latest version of golangci-lint.

Note that the linter reports errors only for code that was changed since previous commit even with the updated configuration. This should allow us to gradually address all linter issues.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,16 +1,84 @@
...
+	cd $(TOOLS_MOD_DIR) && \
+	go build -o $(TOOLS_DIR)/mockery github.com/vektra/mockery/v2
+
+$(TOOLS_DIR)/gofumpt: $(TOOLS_MOD_DIR)/go.mod $(TOOLS_MOD_DIR)/go.sum $(TOOLS_MOD_DIR)/tools.go
+	cd $(TOOLS_MOD_DIR) && \
+	go build -o $(TOOLS_DIR)/gofumpt mvdan.cc/gofumpt
+
+$(TOOLS_DIR)/gci: $(TOOLS_MOD_DIR)/go.mod $(TOOLS_MOD_DIR)/go.sum $(TOOLS_MOD_DIR)/tools.go
+	cd $(TOOLS_MOD_DIR) && \
+	go build -o $(TOOLS_DIR)/gci github.com/daixiang0/gci
```
Comment (haveiss) [inline Makefile:52]: this looks messy, let's refactor this
also add an `install` to rule them all, 

```suggestion
define build_tool
$(TOOLS_DIR)/$(1): $(TOOLS_MOD_DIR)/go.mod $(TOOLS_MOD_DIR)/go.sum $(TOOLS_MOD_DIR)/tools.go
	cd $(TOOLS_MOD_DIR) && \
	go build -o $(TOOLS_DIR)/$(1) $(2)
endef

$(eval $(call build_tool,buf,github.com/bufbuild/buf/cmd/buf))
$(eval $(call build_tool,golangci-lint,github.com/golangci/golangci-lint/cmd/golangci-lint))
$(eval $(call build_tool,mockery,github.com/vektra/mockery/v2))
$(eval $(call build_tool,gofumpt,mvdan.cc/gofumpt))
$(eval $(call build_tool,gci,github.com/daixiang0/gci))

install: $(TOOLS_DIR)/buf $(TOOLS_DIR)/golangci-lint $(TOOLS_DIR)/mockery $(TOOLS_DIR)/gofumpt $(TOOLS_DIR)/gci
	@echo "All tools installed successfully"
```

Comment (sudo-suhas) [reply Makefile:52]: Having individual targets allows us to install each command for required Makefile target. Like `buf` for `make proto`, `golangci-lint` for `make lint` etc. Would want to preserve this property.

Comment (sudo-suhas) [reply Makefile:52]: Do we really need `install` target in this case? Each tool gets installed when required.

Comment (haveiss) [reply Makefile:52]: yes, let's keep individual target,
install really helpful for initial local setup and peace of mind,
so next time we want to run any command like lint or fmt, all the tools are available without worrying about internet

**Review by haveiss — APPROVED** [review]

**Pull request merged**